RenderingNG Fragmentation article (#4555)
GoogleChrome/developer.chrome.com: site/en/articles/renderingng-fragmentation/index.md

---

> layout: 'layouts/blog-post.njk'

Reviewer: Please make sure there is no hero image, like the other RenderingNG articles, and that the social media share metadata indicates Una's picture as the one to go with the link.

title: 'RenderingNG deep-dive: LayoutNG block fragmentation'
description: >

>   In 2021, we will largely complete the process of designing, building and shipping RenderingNG, a next-generation rendering architecture for Chromium.

Reviewer: Needs update.

subhead: >

>   Ready for the next generation of web content

Reviewer: Needs update.

and [BlinkNG](/blog/blinkng/).
{% endAside %}

I'm Morten Stenshorne, a layout engineer on the Blink rendering team at Google. I've been involved in browser engine development since the early 2000s, and I've had a lot of fun, such as helping make the [acid2 test](https://en.wikipedia.org/wiki/Acid2#Timeline_of_passing_applications) pass in the Presto engine (Opera 12 and earlier), and reverse-engineering other browsers to fix table layout in Presto. I've also spent more of those years than I'd like to admit on block fragmentation, and, in particular, [multicol](https://www.w3.org/TR/css-multicol-1/) in Presto, WebKit, and Blink. During the past few years at Google I have mainly been focusing on leading the work of adding block fragmentation support to [LayoutNG](/articles/layoutng/). Join me in this deep-dive into the block fragmentation implementation, as it may very well be the last time that I implement block fragmentation. :)

## What is block fragmentation?

> [Block fragmentation](https://www.w3.org/TR/css-break-3/) is about splitting a CSS block-level box (such as a section or paragraph) into multiple fragments when it doesn't fit as a whole inside one fragment container (called a [_fragmentainer_](https://www.w3.org/TR/css-break-3/#fragmentainer). A fragmentainer is not an element, but represents a column in multi-column layout, or a page in paged media. For fragmentation to happen, the content needs to be inside a fragmentation context. A fragmentation context is most commonly established by a multi-column container (content will be split into columns), or when printing (content will be split into pages). A long paragraph with many lines may need to be split into multiple fragments, so that the first lines are placed in the first fragment, and the remaining lines are placed in subsequent fragments. 

Reviewer: ```suggestion
[Block fragmentation](https://www.w3.org/TR/css-break-3/) is about splitting a CSS block-level box (such as a section or paragraph) into multiple fragments when it doesn't fit as a whole inside one fragment container (called a [_fragmentainer_](https://www.w3.org/TR/css-break-3/#fragmentainer). A fragmentainer is not an element, but represents a column in multi-column layout, or a page in paged media. For fragmentation to happen, the content needs to be inside a *fragmentation context*. A fragmentation context is most commonly established by a multi-column container (content will be split into columns), or when printing (content will be split into pages). A long paragraph with many lines may need to be split into multiple fragments, so that the first lines are placed in the first fragment, and the remaining lines are placed in subsequent fragments. 
```

<figure>
{% Img src="image/kheDArv5csY6rvQUJDbWRscckLr1/cCaTuW0K4aGvX1HHSH6d.png", alt="A paragraph of text broken into two columns.", width="800", height="576" %}

<figcaption>In this example a paragraph has been split into two columns using multi-column layout. Each column is a fragmentainer, representing a fragment of the fragmented flow.</figcaption>
</figure>

> Block fragmentation is analogous to another well-known type of fragmentation: line fragmentation. Any inline element that consists of more than one word (any text node, any `<a>` element, and so on), and allows line breaks, may be split into multiple fragments. Each fragment is placed into a different line box. A line box is the inline fragmentation equivalent to a _fragmentainer_ for columns and pages.

Reviewer: ```suggestion
Block fragmentation is analogous to another well-known type of fragmentation: line fragmentation (otherwise known as "line breaking"). Any inline element that consists of more than one word (any text node, any `<a>` element, and so on), and allows line breaks, may be split into multiple fragments. Each fragment is placed into a different line box. A line box is the inline fragmentation equivalent to a _fragmentainer_ for columns and pages.
```

## What is LayoutNG block fragmentation?

> LayoutNGBlockFragmentation is a rewrite of the fragmentation engine for LayoutNG, and after many years of work, the first parts finally shipped in Chrome 102 earlier this year. This fixed long-standing issues that were essentially unfixable in our "legacy" engine. In terms of data structures, it replaces multiple data structures with _NG fragments_ represented directly in the [fragment tree](/articles/renderingng-data-structures/#the-immutable-fragment-tree).

Reviewer: ```suggestion
LayoutNGBlockFragmentation is a rewrite of the fragmentation engine for LayoutNG, and after many years of work, the first parts finally shipped in Chrome 102 earlier this year. This fixed long-standing issues that were essentially unfixable in our "legacy" engine. In terms of data structures, it replaces multiple pre-NG data structures with _NG fragments_ represented directly in the [fragment tree](/articles/renderingng-data-structures/#the-immutable-fragment-tree).
```

For instance, we now support the ['avoid' value for the 'break-before' and 'break-after' CSS properties](https://www.w3.org/TR/css-break-3/#break-between), which allow authors to avoid breaks right after a header. It generally doesn't look good if the last thing that is placed on a page is a header, while the section's content starts on the next page. Instead, it's better to break **before** the header. See the figure below for an example.

{% Img src="image/kheDArv5csY6rvQUJDbWRscckLr1/bVIi9gBcMJx1pgEGiuxZ.png", alt="The first example shows a heading at the bottom of the page, the second shows it at the top of the following page with its associated content.", width="800", height="589" %}

Chrome 102 also supports fragmentation overflow, so that monolithic (supposed to be unbreakable) content isn't sliced into multiple columns, and paint effects like shadows and transforms are applied correctly.

## Block fragmentation in LayoutNG is now complete

> At the time of writing this, we have completed full block fragmentation support in LayoutNG. Core fragmentation (block containers, including line layout, floats, and out-of-flow positioning) shipped in Chrome 102. Flex and grid fragmentation shipped in Chrome 103, and table fragmentation shipped in Chrome 106. Lastly, [printing](https://bugs.chromium.org/p/chromium/issues/detail?id=1121942) shipped in Chrome 108. Block fragmentation was the last feature that depended on the  legacy engine for performing layout. This means that, as of Chrome 108, the legacy engine will no longer be used to perform layout.

Reviewer: ```suggestion
At the time of writing this, we have completed full block fragmentation support in LayoutNG. Core fragmentation (block containers, including line layout, floats, and out-of-flow positioning) shipped in Chrome 102. Flex and grid fragmentation shipped in Chrome 103, and table fragmentation shipped in Chrome 106. Lastly, [printing](https://bugs.chromium.org/p/chromium/issues/detail?id=1121942) shipped in Chrome 108. Block fragmentation was the last feature that depended on the legacy engine for performing layout. This means that, as of Chrome 108, the legacy engine will no longer be used to perform layout.
```

> In addition to actually laying out the content, LayoutNG supports painting and hit-testing, but we still rely on some legacy data structures for JavaScript APIs that read layout information, such as `offsetLeft` and `offsetTop`.

Reviewer: ```suggestion
In addition to actually laying out the content, LayoutNG data structures support painting and hit-testing, but we still rely on some legacy data structures for JavaScript APIs that read layout information, such as `offsetLeft` and `offsetTop`.
```

In addition, believe it or not, by mid-2019 the majority of the core functionality of LayoutNG block fragmentation layout was already implemented (behind a flag). So, why did it take so long for it to ship? The short answer is: fragmentation has to coexist correctly with various legacy parts of the system, which cannot be removed or upgraded until all dependencies are upgraded. For the long answer, see the following details.

### Legacy engine interaction

Legacy data structures are still in charge of JavaScript APIs that read layout information, so we need to write back data to the legacy engine in a manner that it understands. This includes updating the legacy multi-column data structures, like [LayoutMultiColumnFlowThread](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/layout/layout_multi_column_flow_thread.h;l=144;drc=237a52e7e2b6d22eb1b5912e134384254d5b8974), correctly. 

### Legacy engine fallback detection and handling

> We had to fall back to the legacy layout engine when there was content inside that couldn't yet be handled by LayoutNG block fragmentation. At the time of shipping core LayoutNG block fragmentation (spring 2022), that included flex, grid, tables, and anything that's printed. This was particularly tricky because we need to detect the need for legacy fallback before creating [objects in the layout tree](/articles/layoutng/#a-30000-foot-view-of-layout-engine-architectures). For example, before we know if there's a multi-column container ancestor, and before we know which DOM nodes will become a [formatting context](https://developer.mozilla.org/docs/Web/Guide/CSS/Block_formatting_context) or not. It's a chicken-and-egg problem that doesn't have a perfect solution, but as long as its only misbehavior is false positives (fallback to legacy when there's actually no need), it's ok, because any bugs in that layout behavior are ones Chromium already has, not new ones.

Reviewer: ```suggestion
We had to fall back to the legacy layout engine when there was content inside that couldn't yet be handled by LayoutNG block fragmentation. At the time of shipping core LayoutNG block fragmentation (spring 2022), that included flex, grid, tables, and anything that's printed. This was particularly tricky because we needed to detect the need for legacy fallback before creating [objects in the layout tree](/articles/layoutng/#a-30000-foot-view-of-layout-engine-architectures). For example, we needed to detect before we knew if there was a multi-column container ancestor, and before we knew which DOM nodes would become a [formatting context](https://developer.mozilla.org/docs/Web/Guide/CSS/Block_formatting_context) or not. It's a chicken-and-egg problem that doesn't have a perfect solution, but as long as its only misbehavior is false positives (fallback to legacy when there's actually no need), it's ok, because any bugs in that layout behavior are ones Chromium already has, not new ones.
```

### Pre-paint tree walk

> That's something we [do after layout](/articles/renderingng-architecture/#rendering-pipeline-structure), but before painting. The main challenge is that we still need to walk the [layout object tree](/articles/layoutng/#a-30000-foot-view-of-layout-engine-architectures), but we have NG fragments now—so how do we deal with that? We walk both the layout object and NG fragment trees at the same time! This is quite complicated, because mapping between the two trees isn't trivial. While the layout object tree structure closely resembles that of the DOM tree, the fragment tree is an **output** of layout, not an input to it. Apart from actually reflecting the effect of any fragmentation including inline fragmentation (line fragments) and block fragmentation (column or page fragments), the fragment tree also has a direct parent-child relationship between a [containing block](https://developer.mozilla.org/docs/Web/CSS/Containing_block) and the DOM descendants that have that fragment as their containing block. For example, in the fragment tree, a fragment generated by an absolutely positioned element is a direct child of its containing block fragment, even if there are other nodes in the ancestry chain between the out-of-flow positioned descendant and its containing block.  

Reviewer: ```suggestion
Pre-paint is something we [do after layout](/articles/renderingng-architecture/#rendering-pipeline-structure), but before painting. The main challenge is that we still need to walk the [layout object tree](/articles/layoutng/#a-30000-foot-view-of-layout-engine-architectures), but we have NG fragments now—so how do we deal with that? We walk both the layout object and NG fragment trees at the same time! This is quite complicated, because mapping between the two trees isn't trivial. While the layout object tree structure closely resembles that of the DOM tree, the fragment tree is an **output** of layout, not an input to it. Apart from actually reflecting the effect of any fragmentation including inline fragmentation (line fragments) and block fragmentation (column or page fragments), the fragment tree also has a direct parent-child relationship between a [containing block](https://developer.mozilla.org/docs/Web/CSS/Containing_block) and the DOM descendants that have that fragment as their containing block. For example, in the fragment tree, a fragment generated by an absolutely positioned element is a direct child of its containing block fragment, even if there are other nodes in the ancestry chain between the out-of-flow positioned descendant and its containing block.  
```

## The problems with the legacy fragmentation engine

The legacy engine, designed in an earlier era of the web, doesn't really have a concept of fragmentation, even if fragmentation technically existed back then as well (in order to support printing). Fragmentation support was just something that got bolted on top (printing) or retrofitted (multi-column).

When laying out fragmentable content, the legacy engine lays out everything into a tall strip whose width is the inline-size of a column or page, and height is as tall as it needs to be to contain its content. This tall strip is not rendered to the page—think of it as rendering to a virtual page that is then rearranged for final display. It's conceptually similar to printing an entire paper newspaper article into one column, and then using scissors to cut it into multiple as a second step. (Back in the day, some newspapers actually used techniques similar to this!)

> The legacy engine keeps track of an imaginary page or column boundary in the strip so that it can nudge content that doesn't fit past the boundary, into the next page or column. For example, if only the upper half of a line would fit on what the engine thinks is the current page, it will insert a "pagination strut" to push it down to the position where the engine assumes that the top of the next page is. Then, most of the actual fragmentation work (the "cutting with scissors and placement") takes place after layout during pre-paint and painting, by slicing the tall strip of content into pages or columns (by clipping and translating portions). This made a few things essentially impossible, such as applying transforms and relative positioning **after** fragmentation (which is what the spec requires). Furthermore, while there's some support for table fragmentation in the legacy engine, there's no flex or grid fragmentation support at all.

Reviewer: ```suggestion
The legacy engine keeps track of an imaginary page or column boundary in the strip. That lets it can nudge content that doesn't fit past the boundary into the next page or column. For example, if only the upper half of a line would fit on what the engine thinks is the current page, it will insert a "pagination strut" to push it down to the position where the engine assumes that the top of the next page is. Then, most of the actual fragmentation work (the "cutting with scissors and placement") takes place after layout during pre-paint and painting, by slicing the tall strip of content into pages or columns (by clipping and translating portions). This made a few things essentially impossible, such as applying transforms and relative positioning **after** fragmentation (which is what the spec requires). Furthermore, while there's some support for table fragmentation in the legacy engine, there's no flex or grid fragmentation support at all.
```

Reviewer 2: The grammar became wrong now. "That lets it nudge" WFM.

Here's an illustration of how a three-column layout is represented internally in the legacy engine, before using scissors, placement and glue (we have a specified height, so that only four lines fit, but there's some excess space at the bottom):

{% Img src="image/kheDArv5csY6rvQUJDbWRscckLr1/6pPbkYwDlbboAwLjJti1.png", alt="The internal representation as one column with pagination struts where the content breaks, and the on screen representation as three columns.", width="664", height="411" %}

> Because the legacy layout engine doesn't actually fragment content during layout, there are many strange artifacts, such as relative positioning and transforms not being applied correctly, and box-shadows being clipped at column edges.

Reviewer: ```suggestion
Because the legacy layout engine doesn't actually fragment content during layout, there are many strange artifacts, such as relative positioning and transforms applying incorrectly, and box-shadows being clipped at column edges.
```

> Valid breakpoints generally only occur _between_ siblings (line boxes or blocks), and not, for example,  between a parent and its first child ([class C breakpoints](https://www.w3.org/TR/css-break-3/#end-block) are an exception, but we don't need to discuss those here). There **is** a valid breakpoint for instance before a block sibling with break-before:avoid, but it's somewhere between "perfect" and "last-resort".

Reviewer: ```suggestion
Valid breakpoints generally only occur _between_ siblings (line boxes or blocks), and not, for example, between a parent and its first child ([class C breakpoints](https://www.w3.org/TR/css-break-3/#end-block) are an exception, but we don't need to discuss those here). There **is** a valid breakpoint for instance before a block sibling with break-before:avoid, but it's somewhere between "perfect" and "last-resort".
```
